### PR TITLE
🎈 Integrate usePromiseTransaction with Gnosis Safe

### PR DESCRIPTION
**Pull request by jakvbs**

### CODE REVIEW

**Review by rzadp — COMMENTED** [review]
`safeTransaction` sounds like a secure transaction.
How about `multisigTransaction` ?

File: packages/core/src/hooks/useContractFunction.ts [hunk 1]
```diff
@@ -91,21 +91,31 @@ export function useContractFunction<T extends TypedContract, FN extends Contract
...
         }
         const modifiedArgs = hasOpts ? args.slice(0, args.length - 1) : args
         modifiedArgs.push(modifiedOpts)
 
-        const receipt = await promiseTransaction(contractWithSigner[functionName](...modifiedArgs))
+        let safeTransactionData: string | undefined = undefined
+        try {
+          safeTransactionData = contract.interface.encodeFunctionData(functionName, encodeFunctionDataArgs)
+        } catch (e) {
+          console.error(e)
```
Comment (rzadp) [inline packages/core/src/hooks/useContractFunction.ts:110]: Does it make sense to proceed without `safeTransactionData`?

File: packages/core/src/hooks/useContractFunction.ts [hunk 2]
```diff
@@ -91,21 +91,31 @@ export function useContractFunction<T extends TypedContract, FN extends Contract
...
-          contractWithSigner,
-          functionName,
-          args,
-          gasLimitBufferPercentage
-        )
+        const gasLimit =
+          (await estimateContractFunctionGasLimit(contractWithSigner, functionName, args, gasLimitBufferPercentage)) ??
+          BigNumber.from(0)
 
+        const encodeFunctionDataArgs = args.filter((arg) => !arg.to && !arg.value)
```
Comment (rzadp) [inline packages/core/src/hooks/useContractFunction.ts:98]: Make consistent with `modifiedArgs`

File: packages/core/src/hooks/usePromiseTransaction.ts [hunk 3]
```diff
@@ -52,32 +66,137 @@ export function usePromiseTransaction(chainId: number | undefined, options?: Tra
...
-            ...transaction,
-            chainId: chainId,
-          },
-          submittedAt: Date.now(),
-          transactionName: options?.transactionName,
-        })
-        const receipt = await transaction.wait()
-        setState({ receipt, transaction, status: 'Success', chainId })
-        return receipt
+        if (await isGnosisSafe(chainId, account ?? '')) {
```
Comment (rzadp) [inline packages/core/src/hooks/usePromiseTransaction.ts:90]: Try to extract something.

File: packages/core/src/hooks/useContractFunction.test.tsx [hunk 4]
```diff
@@ -71,13 +71,8 @@ describe('useContractFunction', () => {
     )
 
     await waitForNextUpdate()
-    await result.current.send()
+    await expect(result.current.send()).to.be.rejectedWith('Invalid number of arguments.')
```
Comment (rzadp) [inline packages/core/src/hooks/useContractFunction.test.tsx:74]:
```suggestion
    await expect(result.current.send()).to.be.rejectedWith('Invalid number of arguments for function "approve". Expected: 2. Received: 0.')
```

File: packages/core/src/helpers/gnosisSafeUtils.ts [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+
+export const calculateSafeTransactionHash = (
+  safe: Contract,
+  safeTx: SafeTransaction,
+  chainId: BigNumberish
+): string => {
+  return utils._TypedDataEncoder.hash({ verifyingContract: safe.address, chainId }, EIP712_SAFE_TX_TYPE, safeTx)
+}
+
+export const getLatestNonce = async (chainId: number, safeAddress: string): Promise<number | null | undefined> => {
```
Comment (rzadp) [inline packages/core/src/helpers/gnosisSafeUtils.ts:76]: Add jsdoc explaining the difference between `null` and `undefined`.

File: packages/core/src/helpers/gnosisSafeUtils.ts [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        contract.removeListener('ExecutionSuccess', onExecutionSuccess)
+
+        const transaction = await event.getTransaction()
+        const receipt = await event.getTransactionReceipt()
+
+        resolve({ transaction, receipt, rejected: false })
+      } else {
+        const currentNonce = await contract.nonce()
+
+        if (Number(currentNonce) > Number(safeTx.nonce)) {
```
Comment (rzadp) [inline packages/core/src/helpers/gnosisSafeUtils.ts:124]: Add some explanation here, with some domain info about gnosis.

File: packages/core/src/hooks/usePromiseTransaction.ts [hunk 7]
```diff
@@ -45,39 +53,49 @@ export async function estimateContractFunctionGasLimit(
...
+
 const isDroppedAndReplaced = (e: any) =>
   e?.code === errors.TRANSACTION_REPLACED && e?.replacement && (e?.reason === 'repriced' || e?.cancelled === false)
 
 export function usePromiseTransaction(chainId: number | undefined, options?: TransactionOptions) {
   const [state, setState] = useState<TransactionStatus>({ status: 'None' })
   const { addTransaction } = useTransactionsContext()
   const { addNotification } = useNotificationsContext()
+  const { library, account } = useEthers()
+  let gnosisSafeContract: Contract | undefined = undefined
```
Comment (rzadp) [inline packages/core/src/hooks/usePromiseTransaction.ts:75]: const gnosisSafeContract = useGnosisSafeContract(account, library)

**Review by rzadp — APPROVED** [review]

**Comment by loucaponi** [comment]
Thanks for making this change, we are looking forward to integrating it into our project.  Was curious if there is an estimated date for the next release? Thanks.
cc: @jakvbs @rzadp

**Comment by rzadp** [comment]
We want to make a release today.

In any event, you can always use a `dev` version of `@usedapp/core` - it is published to NPM on every commit to master.

**Comment by loucaponi** [comment]
@rzadp Much appreciated 🙏🏼

**Pull request merged**